### PR TITLE
Custom DPO losses support

### PR DESCRIPTION
**Pull request by krammnic**
#### Context
What is the purpose of this PR? Is it to
- [X] add a new feature
- [ ] fix a bug
- [ ] update tests and/or documentation
- [ ] other (please add here)

Please link to any issues this PR addresses.

#### Changelog
What are the changes made in this PR?
* #2292 

#### Test plan
Please make sure to do each of the following if applicable to your PR. If you're unsure about any one of these just ask and we will happily help. We also have a [contributing page](https://github.com/pytorch/torchtune/blob/main/CONTRIBUTING.md) for some guidance on contributing.

- [X] run pre-commit hooks and linters (make sure you've first installed via `pre-commit install`)
- [ ] add [unit tests](https://github.com/pytorch/torchtune/tree/main/tests/torchtune) for any new functionality
- [ ] update [docstrings](https://github.com/pytorch/torchtune/tree/main/docs/source) for any new or updated methods or classes
- [ ] run unit tests via `pytest tests`
- [ ] run recipe tests via `pytest tests -m integration_test`
- [ ] manually run any new or modified recipes with sufficient proof of correctness
- [ ] include relevant commands and any other artifacts in this summary (pastes of loss curves, eval results, etc.)

#### UX
If your function changed a public API, please add a dummy example of what the user experience will look like when calling it.
Here is a [docstring example](https://github.com/pytorch/torchtune/blob/6a7951f1cdd0b56a9746ef5935106989415f50e3/torchtune/modules/vision_transformer.py#L285)
and a [tutorial example](https://pytorch.org/torchtune/main/tutorials/qat_finetune.html#applying-qat-to-llama3-models)

- [ ] I did not change any public API
- [X] I have added an example to docs or docstrings

### CODE REVIEW

**Comment by krammnic** [comment]
Pushed small fixes and I was able to forward!

File: torchtune/rlhf/loss/dpo.py [hunk 1]
```diff
@@ -85,15 +80,15 @@ def forward(
...
         )
         rejected_rewards = (
-            self.beta * (policy_rejected_logps - reference_rejected_logps).detach()
+            self.beta * (policy_inputs.rejected_logps - reference_inputs.rejected_logps).detach()
         )
 
         return losses, chosen_rewards, rejected_rewards
 
-
+@deprecated(msg="RSOLoss will be deprecated in an upcoming release.")
```
Comment (ebsmothers) [inline torchtune/rlhf/loss/dpo.py:91]: Why are we marking as deprecated? Seems this will not be usable with the current changes anyways?

File: recipes/lora_dpo_single_device.py [hunk 2]
```diff
@@ -533,31 +536,25 @@ def train(self) -> None:
...
-                    policy_chosen_log_probs,
-                    policy_rejected_log_probs,
-                    policy_chosen_logits,
-                    policy_rejected_logits,
-                ) = self.concatenated_forward(self._model, batch)
+                policy_chosen_rejected_outputs = self.concatenated_forward(
+                    self._model, batch
+                )
+                
+                print("forwared")
```
Comment (ebsmothers) [inline recipes/lora_dpo_single_device.py:543]: remove

**Review by ebsmothers — COMMENTED** [review]
One question and one nit. After that we're good to go!

File: recipes/lora_dpo_single_device.py (hunk 2 above)
Comment (krammnic) [reply recipes/lora_dpo_single_device.py:543]: Oops

**Review by ebsmothers — APPROVED** [review]
Thanks for patiently addressing all the comments. This looks good to me!
